plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return fmt.Errorf("unable to start database transaction: %w", err)
+	}
+	defer func() {
+		// this always fails if tx.Commit() was called before, ignore error
+		_ = tx.Rollback()
+	}()
+
+	// clean expired locks first, this simplifies the logic later
+	if err := d.cleanExpired(tx, targets, time.Now()); err != nil {
+		return err
+	}
+
+	locks, err := d.queryLocks(tx, targets)
+	if err != nil {
+		return err
+	}
+
+	// detect conflicts (unlock on foreign locks) and warn about them,
+	// but don't abort.
+	conflicts := make([]dblock, 0)
+	for _, lock := range locks {
+		if lock.jobID != jobID {
+			conflicts = append(conflicts, lock)
+		}
+	}
+	if len(conflicts) > 0 {
+		log.Warningf("unable to unlock some targets due to different lock owners: %+v", conflicts)
+	}
+
+	// only remove locks held by the owner
+	del := "DELETE FROM locks WHERE job_id = ? AND target_id IN " + listQueryString(len(targets)) + ";"
+	queryList := make([]interface{}, 0, len(targets)+1)
+	queryList = append(queryList, jobID)
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+	_, err = tx.Exec(del, queryList...)
+	if err != nil {
+		return fmt.Errorf("unable to unlock targets: %w", err)
+	}
+
+	return tx.Commit()
+}
+
+// Lock locks the given targets.
+// See target.Locker for API details
+func (d *DBLocker) Lock(jobID types.JobID, targets []*target.Target) error {
+	if jobID == 0 {
+		return fmt.Errorf("invalid lock request, jobID cannot be zero (targets: %v)", targets)
+	}
+	log.Debugf("Requested to lock %d targets for job ID %d: %v", len(targets), jobID, targets)
+
+	return d.handleLock(int64(jobID), targetIDList(targets), d.lockTimeout)
+}
+
+// Unlock unlocks the given targets.
+// See target.Locker for API details
+func (d *DBLocker) Unlock(jobID types.JobID, targets []*target.Target) error {
+	if jobID == 0 {
+		return fmt.Errorf("invalid unlock request, jobID cannot be zero (targets: %v)", targets)
+	}
+	log.Debugf("Requested to unlock %d targets for job ID %d: %v", len(targets), jobID, targets)
+
+	return d.handleUnlock(int64(jobID), targetIDList(targets))
+}
+
+// RefreshLocks refreshes (or locks!) the given targets.
+// See target.Locker for API details
+func (d *DBLocker) RefreshLocks(jobID types.JobID, targets []*target.Target) error {
+	if jobID == 0 {
+		return fmt.Errorf("invalid refresh request, jobID cannot be zero (targets: %v)", targets)
+	}
+	log.Debugf("Requested to refresh %d targets for job ID %d: %v", len(targets), jobID, targets)
+
+	return d.handleLock(int64(jobID), targetIDList(targets), d.refreshTimeout)
+}
+
+// ResetAllLocks resets the database and clears all locks, regardless of who owns them.
+// This is primarily for testing, and should not be used by used in prod, this
+// is why it is not exposed by target.Locker
+func (d *DBLocker) ResetAllLocks() error {
```

Comment:
being internal, unexport it?

Reply:
No then the semi-external integration tests can't reach it :/